README.md
# PiView

A Raspberry Pi system information package.

![PiView Icon](PiView-0.25x.png)

PiView provides the details of the Raspberry Pi currently being interrogated. System information includes, but is not limited to:

- **CPU**: max load across cores, temperature, clock speed
- **GPU**: temperature
- **HARDWARE**: bluetooth, i2c, spi, camera statuses
- **HOST**: boot time, model, name, revision, serial number, uptime
- **NETWORK**: host name, interface names, ip addresses, mac addresses
- **STORAGE**: total disk capacity, free disk capacity, total RAM and free RAM

Also includes a small utility library with:

- conversion of bytes into Kilobytes, Megabytes, Gigabytes and up
- create list with a quartet of integer numbers representing the IPv4 Address

## Changes

See the [CHANGES](CHANGES.md) document for details of updates and changes.

## Requirements

This project requires a number of packages, including:

- psutils

Remaining packages are Python 'built-ins'.

## Building

Building is in 3 stages:

1. Build package
2. Update and build docs
3. Publish package

### Build package

```shell
python3 -m build
```

### Publish package

```shell
python3 -m twine upload --repository PiView-AG dist/*
```

### Build Docs

If you wish to build the docs from scratch, then remove the docs/source/modules.rst and the docs/source/PiView_AG.rst files before using the build steps.

#### First Build
From there, first time through use:

```shell
cd docs
sphinx-apidoc -o source ../PiView_AG -a -f 
sphinx-build -b html source html -a -j 2
```

#### Re-Builds
Subsequent runs, if you are in the documentation (docs) folder already:

```shell
sphinx-apidoc -o source ../PiView_AG -a -f 
make clean && make html
```

## Acknowledgements

A very large thank you to Matt Hawkins upon whose code this package is based: [https://www.raspberrypi-spy.co.uk/](https://www.raspberrypi-spy.co.uk/).

The original code may be found as [mypi.py](https://github.com/tdamdouni/Raspberry-Pi-DIY-Projects/blob/master/MattHawkinsUK-rpispy-misc/python/mypi.py).

## About the Author

Adrian Gould has been coding for over 40 years, starting his coding in Sinclair ZX-80 Basic and Machine Code, through Pascal, Modula-2, Occam, Prolog and many others to the current swathe of Python, C#, PHP, JS, and other languages today.

He believes that it is a continuous process to learn any (coding) language, and will always say he is not an expert.

He is a full time educator who lives and works in Perth, Western Australia.

## Copyright

Copyright Adrian Gould, 2021-. Licensed under
the [Open Software License version 3.0](./LICENSE.txt)


CHANGES.md
# Changes to PiView-AG
## TO DO
This is a list of components to be implemented:

- **STORAGE**: for all storage devices: disk capacity, free disk capacity
- **CPU**: load across each core of the processor (as a list/dictionary)

Future features will be added to this list as required.

## COMPLETE
These items are completed.

### 1.0.0 Initial Release
Some enhancements to come - such as return all the attached storage statistics

Added:
- python version (HOST)
- camera supported / detected (HARDWARE)

### 0.5.0
Added the following:
- ram total and free
- storage total and free for 'all disks' in total
- name to Host, this is a temporary version until further investigation done, use the host 
  name method in the network section to get host name

To do:
- statistics (total space, free) for each attached storage device


### 0.4.0 Network Features
The following are implemented in this version:
- host name
- interface names
- ip addresses
- mac addresses

Fixed missing self references in classes, removed get_ from function names
Added missing file headers...

### 0.3.0 Host Features
The following have been implemented:
- boot time 
- model
- serial number
- uptime
- revision


- ### 0.2.0 Hardware Features
Added Hardware checking for:
	- SPI
	- I2C
	- BT
Updated [[README]]
Design and added PiView Icon

- ### 0.1.1 GPU Features
Added:
	- GPU temperature

- ### 0.1.0 CPU Features
Added:
	- max load across cores
	- processor temperature
	- processor clock speed

- ### 0.0.3 Setup fixes
Small fixes to setup.cfg, and a source reformat.

- ### 0.0.2 Utils
Added Utils to the package. Utils includes:
	- format_bytes
	- draw_line

- ### 0.0.1 Initial Version
Blank project, containing:
	- starter folder structure
	- [[README]]  
	- [[CHANGES]]
	- [[LICENSE.txt]]

tests/test_Utils_line_draw.py
from PiView_AG.Utils import Utils

def test_draw_no_line():
    utils = Utils()
    results = utils.draw_line(length=0)
    expected = ""
    assert results == expected


def test_draw_no_characters():
    utils = Utils()
    results = utils.draw_line(characters="1234567890", length=0)
    expected = ""
    assert results == expected


def test_draw_line_ten_spaces():
    utils = Utils()
    results = utils.draw_line(characters=" ", length=10)
    expected = "          "
    assert results == expected


def test_draw_line_ten_dashes():
    utils = Utils()
    results = utils.draw_line(characters="-", length=10)
    expected = "----------"
    assert results == expected


def test_draw_line_eleven_dash_dots():
    utils = Utils()
    results = utils.draw_line(characters="-.", length=11)
    expected = "-.-.-.-.-.-"
    assert results == expected


def test_draw_line_two_characters_from_six():
    utils = Utils()
    results = utils.draw_line(characters="123456", length=2)
    expected = "12"
    assert results == expected


def test_draw_default_line():
    utils = Utils()
    results = utils.draw_line()
    expected = "-=--=--=--=--=--=--=--=--=--=--=--=--=--"
    assert results == expected


tests/test_Utils_format_bytes.py
from PiView_AG.Utils import Utils

short_labels = {0: '', 1: 'K', 2: 'M', 3: 'G', 4: 'T',
                5: 'P', 6: 'E', 7: 'Y'}
long_labels = {0: ' ', 1: ' Kilo', 2: ' Mega', 3: ' Giga',
               4: ' Tera', 5: ' Peta', 6: ' Exa', 7: ' Yotta'}


def test_format_as_bytes():
    utils = Utils()
    results = utils.format_bytes(size=1023, style='short')
    expected = (1023, 'B')
    assert results == expected


def test_format_as_short_kilobytes():
    utils = Utils()
    results = utils.format_bytes(size=765.4*1024, style='short')
    expected = (765.4, 'KB')
    assert results == expected


def test_format_as_short_megabytes():
    utils = Utils()
    results = utils.format_bytes(size=1.5*1024**2, style='short')
    expected = (1.5, 'MB')
    assert results == expected


def test_format_as_short_megabytes_2():
    utils = Utils()
    results = utils.format_bytes(size=9753*1024, style='short')
    expected = (9.5244140625, 'MB')
    assert results == expected


def test_format_as_short_gigabytes():
    utils = Utils()
    results = utils.format_bytes(size=3*1024**3, style='short')
    expected = (3.0, 'GB')
    assert results == expected


def test_format_as_short_terabytes():
    utils = Utils()
    results = utils.format_bytes(size=723*1024**4, style='short')
    expected = (723.0, 'TB')
    assert results == expected


def test_format_as_short_petabytes():
    utils = Utils()
    results = utils.format_bytes(size=1024**5, style='short')
    expected = (1.0, 'PB')
    assert results == expected


def test_format_as_short_exabytes():
    utils = Utils()
    results = utils.format_bytes(size=1024**6, style='short')
    expected = (1.0, 'EB')
    assert results == expected


def test_format_as_short_yottabytes():
    utils = Utils()
    results = utils.format_bytes(size=1024**7, style='short')
    expected = (1.0, 'YB')
    assert results == expected




def test_format_as_bytes():
    utils = Utils()
    results = utils.format_bytes(size=1023, style='long')
    expected = (1023, 'bytes')
    assert results == expected


def test_format_as_long_kilobytes():
    utils = Utils()
    results = utils.format_bytes(size=765.4*1024, style='long')
    expected = (765.4, 'Kilobytes')
    assert results == expected


def test_format_as_long_megabytes():
    utils = Utils()
    results = utils.format_bytes(size=1.5*1024**2, style='long')
    expected = (1.5, 'Megabytes')
    assert results == expected


def test_format_as_long_megabytes_2():
    utils = Utils()
    results = utils.format_bytes(size=9753*1024, style='long')
    expected = (9.5244140625, 'Megabytes')
    assert results == expected


def test_format_as_long_gigabytes():
    utils = Utils()
    results = utils.format_bytes(size=3*1024**3, style='long')
    expected = (3.0, 'Gigabytes')
    assert results == expected


def test_format_as_long_terabytes():
    utils = Utils()
    results = utils.format_bytes(size=723*1024**4, style='long')
    expected = (723.0, 'Terabytes')
    assert results == expected


def test_format_as_long_petabytes():
    utils = Utils()
    results = utils.format_bytes(size=1024**5, style='long')
    expected = (1.0, 'Petabytes')
    assert results == expected


def test_format_as_long_exabytes():
    utils = Utils()
    results = utils.format_bytes(size=1024**6, style='long')
    expected = (1.0, 'Exabytes')
    assert results == expected


def test_format_as_long_yottabytes():
    utils = Utils()
    results = utils.format_bytes(size=1024**7, style='long')
    expected = (1.0, 'Yottabytes')
    assert results == expected

PiView_AG/CPU.py
# Project:    PiView
# Filename:   CPU.py
# Location:   ./PiView_AG
# Author:     Adrian Gould <adrian.gould@nmtafe.wa.edu.au>
# Created:    10/04/21
#
# This file provides the following features, methods and associated supporting
# code:
# - speed
# - maximum load
# - temperature

import subprocess

import psutil

from PiView_AG import Utils


class CPU:
    def speed(self):
        """
        Get the CPU frequency using the vcgencmd on Linux based systems

        :rtype: integer
        :return: The CPU frequency in MHz
        """
        try:
            output = subprocess.check_output(
                ['vcgencmd', 'get_config', 'arm_freq'])
            output = output.decode()
            lines = output.splitlines()
            line = lines[0]
            freq = line.split('=')
            freq = freq[1]
        except:
            freq = '0'
        return freq

    def max_load(self):
        """
        This function returns the maximum "CPU load" across all CPU cores,
        or a random value if the actual CPU load can't be determined.

        :rtype: float
        :return: The maximum CPU Load in range 0-100
        """
        if psutil is not None:
            return max(psutil.cpu_percent(percpu=True))
        else:
            return Utils.random_percentage()

    def temperature(self):
        """
        Requests the CPU temperature from the vcgencmd returning the
        result to the caller as a string with a floating point value to 2DP

        :rtype: float
        :return: The CPU temperature in Celcius
        """
        try:
            temp = subprocess.check_output(['vcgencmd', 'measure_temp'])
            temp = temp[5:-3]
        except:
            temp = '0.0'
        temp = round(temp, 2)
        return temp


PiView_AG/Host.py
# Project:    PiView
# Filename:   Host.py
# Location:   ./PiView_AG
# Author:     Adrian Gould <adrian.gould@nmtafe.wa.edu.au>
# Created:    10/04/21
#
# This file provides the following features, methods and associated supporting
# code:
# - get boot time
# - get model
# - get revision
# - get serial
# - get uptime
#
# The following may be deprecated at some point:
# - name

from datetime import datetime, timedelta
from socket import gethostname
from sys import platform

import psutil


class Host():

    def boot_time(self):
        """
        Determines the time the device was started

        :rtype: datetime
        :return: How long ago the Pi was booted
        """
        booted_at = datetime.fromtimestamp(psutil.boot_time())
        return booted_at

    def model(self):
        """
        Provide Pi Model Details

        Extracts the details from the device tree model file

        :rtype: string
        :return: The model name and other identifying details
        """
        try:
            my_model = open('/proc/device-tree/model').readline()
        except:
            my_model = "Error"

        return my_model

    def name(self):
        """
        Provides the host name to the user

        :rtype: string
        :return: The host name of the Pi
        """
        return gethostname()

    def python(self):
        """
        Get current Python version

        :rtype: string
        :return: A string containing the version of python that is being used
        """
        pythonv = platform.python_version()
        return pythonv

    def revision(self):
        """
        Provide board revision details

        The details are extracted from the cpu info file

        :rtype: string
        :return: The revision number of the Pi motherboard
        """
        my_revision = "Error"
        try:
            f = open('/proc/cpuinfo', 'r')
            for line in f:
                if line[0:8] == 'Revision':
                    my_revision = line[11:-1]
            f.close()
        except:
            my_revision = "Error"

        return my_revision

    def serial(self):
        """
        Provide the Serial Number of the Pi CPU

        The details are extracted from the cpu info file

        :rtype: string
        :return: The Pi's serial number
        """
        my_cpu_serial = "Error"
        try:
            f = open('/proc/cpuinfo', 'r')
            for line in f:
                if line[0:6] == 'Serial':
                    my_cpu_serial = line[10:26]
            f.close()
        except:
            my_cpu_serial = "Error"

        return my_cpu_serial

    def uptime(self):
        """
        Determines the amount of time the device has been running for in
        seconds

        :rtype: float
        :return: The time that the Pi has been 'up' for
        """
        booted_at = self.boot_time()
        current_at = datetime.now()
        uptime_seconds = (current_at - booted_at).total_seconds()
        uptime = timedelta(seconds=uptime_seconds)
        return uptime


PiView_AG/Network.py
# Project:    PiView
# Filename:   Network.py
# Location:   ./PiView_AG
# Author:     Adrian Gould <adrian.gould@nmtafe.wa.edu.au>
# Created:    10/04/21
#
# This file provides the following features, methods and associated
# supporting code:
# - host name
# - interface names
# - ip addresses
# - mac addresses

import os
from socket import gethostname


class Network:

    def host_name(self):
        """Provide the host name to the user

        :rtype: string
        :return: The host name of the Pi
        """
        # return platform.node()
        # return platform.uname()[1]
        return gethostname()

    def eth_name(self, type=None):
        """Provide the Ethernet interface name

        :param type: string, possible options are: enx or eth
        :rtype: string
        :return: The network interface name
        """
        options = ['enx', 'eth']
        interface = None
        if type == 'w':
            options = ['wla']
        if type == 'l':
            options = ['lo']
        try:
            for root, dirs, files in os.walk('/sys/class/net'):
                for dir in dirs:
                    for option in options:
                        if dir[:3] == option:
                            interface = dir
        except:
            interface = "None"
        return interface

    def mac(self, interface='eth0'):
        """Provides the hardware MAC address for the interface requested

        Default is eth0

        :param interface: string, the interface name to query
        :rtype: string
        :return: Mac address of the selected interface
        """
        # Return the MAC address of named Ethernet interface
        try:
            line = open('/sys/class/net/%s/address' % interface).read()
        except:
            line = "None"
        return line[0:17]

    def ip(self, interface='eth0'):
        """Provide IP Address from the named interface

        Default is eth0

        Uses the ifconfig command to create a text file, then processes this
        file to obtain the IP address.

        :param interface: string, the interface to obtain the IP address for
        :rtype: string
        :return: IPv4 address of the selected interface
        """
        try:
            filename = 'ifconfig_' + interface + '.txt'
            os.system('ifconfig ' + interface + ' > /home/pi/' + filename)
            f = open('/home/pi/' + filename, 'r')
            skip_line = f.readline()  # skip 1st line
            line = f.readline()  # read 2nd line
            line = line.strip()
            f.close()

            if line.startswith('inet '):
                a, b, c = line.partition('inet ')
                a, b, c = c.partition(' ')
                a = a.replace('addr:', '')
            else:
                a = 'None'

            return a

        except:
            return 'Error'


PiView_AG/Storage.py
# Project:    PiView
# Filename:   Storage.py
# Location:   ./PiView_AG
# Author:     Adrian Gould <adrian.gould@nmtafe.wa.edu.au>
# Created:    10/04/21
#
# This file provides the following features, methods and associated
# supporting code:
# - total ram and free ram (tuple)
# - total disk storage and free storage space (tuple)
# - disc storage for all disks connected to Pi (Total and Free) as a list or
# dictionary (To be determined)

import subprocess


class Storage():
    def ram(self):
        """
        Provide the total RAM and free RAM to the user as a tuple

        :rtype: tuple
        :return: (Total, Free)

        """
        output = subprocess.check_output(['free', '-m'])
        lines = output.splitlines()
        line = str(lines[1])
        ram = line.split()
        # total/free
        return ram[1], ram[3]

    def disc(self):
        """
        Provide the total disc space and the disc space that is free

        :rtype: tuple
        :return: (Total, Free)

        """
        output = subprocess.check_output(['df', '-h'])
        lines = output.splitlines()
        line = str(lines[1])
        disk = line.split()
        # total/free
        return disk[1], disk[3]

    def all_discs(self):
        """
        Provide the user with the storage space (Total, Free) for each
        disc attached to the Pi as a dictionary.

        The dictionary will have: "disc name" : (total storage, free storage)
        for each disc.

        :return disc stats: dictionary of tuples
        """
        disc_stats = {}
        # For each disc:
        #   obtain free space and total space
        #   add "disk_name" :(total, free) to 'disc_stats' dictionary
        return disc_stats

    def disk(self):
        """
        Alias for disc()

        see :func:`~PiView_AG.Storage.disc`
        """
        return self.disc()

    def all_disks(self):
        """
        Alias for all_discs()

        see :func:`~PiView_AG.Storage.all_discs`
        """
        return self.all_discs()


PiView_AG/Hardware.py
# Project:    PiView
# Filename:   Hardware.py
# Location:   ./PiView_AG
# Author:     Adrian Gould <adrian.gould@nmtafe.wa.edu.au>
# Created:    10/04/21
#
# This file provides the following features, methods and associated supporting
# code:
# - bt (Bluetooth status)
# - camera (camera status)
# - spi (SPI Bus status)
# - i2c (I2C status)

import subprocess


class Hardware:

    def bt(self):
        """
        Check if Bluetooth module is enabled

        :rtype: boolean
        :return: True|False
        """
        bt = False
        try:
            c = subprocess.Popen("lsmod", stdout=subprocess.PIPE)
            gr = subprocess.Popen(["grep", "bluetooth"], stdin=c.stdout,
                                  stdout=subprocess.PIPE)
            output = gr.communicate()[0]
            if output[:9] == 'bluetooth':
                bt = True
        except:
            pass
        return bt

    def spi(self):
        """
        Check if SPI bus is enabled by checking for spi_bcm2 modules

        :rtype: boolean
        :return: True|False
        """
        spi = False
        try:
            c = subprocess.Popen("lsmod", stdout=subprocess.PIPE)
            gr = subprocess.Popen(["grep", "spi_bcm2"], stdin=c.stdout,
                                  stdout=subprocess.PIPE)
            output = gr.communicate()[0]
            if output[:8] == 'spi_bcm2':
                spi = True
        except:
            pass
        return spi

    def i2c(self):
        """
        Check if I2C bus is enabled by checking for i2c_bcm2 modules

        :rtype: boolean
        :return: True|False
        """
        i2c = False
        try:
            c = subprocess.Popen("lsmod", stdout=subprocess.PIPE)
            gr = subprocess.Popen(["grep", "i2c_bcm2"], stdin=c.stdout,
                                  stdout=subprocess.PIPE)
            output = gr.communicate()[0]
            if output[:8] == 'i2c_bcm2':
                i2c = True
        except:
            pass
        return i2c

    def camera(self):
        """
        Check if camera is enabled and present

        :rtype: dictionary
        :return: Details in form {"supported": boolean, "detected": boolean}
        """
        camera = {"supported": None, "detected": None}
        try:
            c = subprocess.Popen(["/opt/vc/bin/vcgencmd", "get_camera"],
                                 stdout=subprocess.PIPE)
            output = c.communicate()[0]
            supported = output[10] == ord("1")
            detected = output[21] == ord("1")
            camera = {"supported": supported, "detected": detected}
        except:
            pass
        return camera


PiView_AG/Utils.py
# Project:    PiView
# Filename:   Utils.py
# Location:   ./PiView_AG
# Author:     Adrian Gould <adrian.gould@nmtafe.wa.edu.au>
# Created:    10/04/21
#
# This file provides the following features, methods and associated
# supporting code:
# - draw a line using a provided string
# - format a number (bytes) into Bytes, KB, MB, GB,...
# - random number (percentage) between max and min percentages

import random


class Utils:

    def __init__(self):
        return None

    def draw_line(self, characters="-=-", length=40):
        """
        Draw a line of characters using a given character and length

        :param characters: string, the character(s) to draw with
        :param length: integer), the length of the line
        :rtype: string
        :return: A string of exactly length characters
        """
        characters_length = len(characters)
        repeats = length // characters_length
        extras = length % characters_length

        line = characters * repeats

        for count in range(extras):
            next_character = characters[count]
            line = f"{line}{next_character}"

        return line

    # Original found in:
    # https://stackoverflow.com/questions/12523586/python-format-size
    # -application
    # -converting-b-to-kb-mb-gb-tb/37423778
    def format_bytes(self, size=0, style=None):
        """
        Formats the given value into Bytes, Kilobytes, Megabytes, ...

        Using Byte shorthand by default - B, KB, MB, ...

        The style may be:

        - None | short | s  -- Short labels
        - long | l          -- Long labels

        If style is anything other than above, then defaults to long format.

        :param size: integer, defaults to 0
        :param style: string, defaults to None
        :rtype: tuple
        :return: (float, string)
        """
        power = 2 ** 10  # 2**10 = 1024
        n = 0
        short_labels = {0: '', 1: 'K', 2: 'M', 3: 'G', 4: 'T',
                        5: 'P', 6: 'E', 7: 'Y'}
        long_labels = {0: '', 1: 'Kilo', 2: 'Mega', 3: 'Giga',
                       4: 'Tera', 5: 'Peta', 6: 'Exa', 7: 'Yotta'}
        short_end = "B"
        long_end = "bytes"
        while size >= power:
            size /= power
            n += 1

        if style in [None, "short", 's']:
            power_labels = short_labels.copy()
            suffix = short_end
        else:
            power_labels = long_labels.copy()
            suffix = long_end

        return size, power_labels[n] + suffix

    def random_percentage(self, min_percentage=0, max_percentage=100):
        """
        This function returns a random percentage.
        Useful for simulations when developing monitoring dashboards


        :param min_percentage: float, minimum value to return, default 0.0
        :param max_percentage: float, maximum to return, default 100.0
        :rtype: float
        :return: A random CPU load value between 0% and 100% to 1DP
        """
        load = random.gauss(55, 10)
        if load < min_percentage:
            return 0.0
        elif load > max_percentage:
            return 100.0
        else:
            return round(load, 1)
